File: Rust/src/binary_search.rs
pub fn search(nums: Vec<i32>, target: i32) -> i32 {
    use std::cmp::Ordering;
    let (mut left, mut right) = (0, nums.len() - 1);
    while left <= right {
        let mid = left + (right - left) / 2;
        match nums[mid].cmp(&target) {
            Ordering::Equal => return mid as i32,
            Ordering::Greater => right = mid - 1,
            Ordering::Less => left = mid + 1,
        }
    }
    -1
}

File: C++/src/binary_search.cpp
#include <binary_search.hpp>
#include <vector>
using namespace std;

int Solution::search(vector<int> &nums, int target) {
  int start = 0;
  int end = nums.size() - 1;

  while (start <= end) {
    int middle = (end - start) / 2 + start;

    if (nums[middle] == target) {
      return middle;
    }

    if (nums[middle] < target) {
      start = middle + 1;
    } else {
      end = middle - 1;
    }
  }

  return -1;
}